padl/transforms.py

```diff
@@ -1743,10 +1861,56 @@ def __call__(self, args):
         :param args: Arguments to call with.
         :return: Output from series of transforms.
         """
-        for transform_ in self.transforms:
-            args = transform_.pd_call_transform(args)
+        _in_args = args
+        for i, transform_ in enumerate(self.transforms):
+            try:
+                args = transform_.pd_call_transform(args)
+            except Exception as err:
+                self._pd_trace_error(i, _in_args)
+                raise err
         return args
 
+    def _pd_get_non_target_stage_idx(self, stage_used):
+        return 1 if self._pd_group else len(self)
+
+    def _pd_get_target_stage_idx(self, is_entire_transform: bool):
+        return 0 if self._pd_group and not is_entire_transform else _pd_trace[-1].error_position
+
+    def _pd_get_error_idx(self, stage: str):
+        """Track the index where a :class:`Compose` fails from the one that fails
+        on :meth:`self.pd_forward`.
+
+        Examples:
+            t = t_1 >> t_2 >> batch >> t_3 >> t_4
+            then,
+            t.pd_forward = t_3 >> t_4.
+            If we know that :meth:`t.pd_forward` is failing on `t_4`, which is its element 1, then
+            `t` is failing on len(t.pd_preprocess) + 1.
+
+            t = t_1 >> t_2 >> batch >> ((t_3 >> t_4) + (t_5 >> t_6))
+            then,
+            t.pd_forward = (t_3 >> t_4) + (t_5 >> t_6).
+            No matter what branch is failing on :meth:`t.pd_forward`, the error on `t` is on
+            len(t.pd_preprocess) + 0.
+        """
+        assert stage in ('forward', 'postprocess')
+        preprocess = self.pd_preprocess
+        forward = self.pd_forward
+        postprocess = self.pd_postprocess
+        preprocess_used = False if (isinstance(preprocess, Identity)) and \
+                                   not isinstance(self.__getitem__(0), Identity) else True
+
+        preprocess_idx = preprocess._pd_get_non_target_stage_idx(preprocess_used)
+        if stage == 'forward':
+            is_entire_transform = str(self) == str(forward)
```

Comment:
I feel like comparing the strings of the transforms to check they are the same is very fragile. 

 We should check with @wuhu if we can reimplement the comparison operator for two transforms.

Reply:
got you, but it is normally the way we do it. but it would be nice to have a stronger way yeah

Reply:
I see, even if it is the way we do it now, maybe it's worth putting this into the overloaded comparison operator `==` or as a separate function. This will make it easier to update in the future.

Reply:
should I write the overloaded comparison now?

Reply:
@wuhu What do you think? Should we overload the comparison operator for transforms?

Reply:
done

Reply:
For some reason creating the following method:
```
def __eq__(self, other):
    return str(self) == str(other)
```
is causing error in pytorch `TypeErrors` like:
```
TypeError: unhashable type: '<name_of_the_class>`
```
when excecuting statements like: 
```
memo = set()
if self not in memo:
    ....
```

Reply:
Could you provide a minimal code example to reproduce the error?

Reply:
yes, adding the previous `__eq__` method to `Transform`, and doing, for example:

```
@transform
def to_tensor(x):
    return torch.LongTensor(x)

emb = transform(torch.nn.Embedding)(10, 8)

t = to_tensor >> batch >> emb

list(t.train_apply([[9, 8, 8], [4, 2, 2], [5, 9, 1], [3, 7, 6]], batch_size=2, num_workers=0))
```
or also running the tests after adding `__eq__`. I think it is getting in conflict with other `__eq__` methods.

Reply:
yes, adding the previous `__eq__` method to `Transform`, and doing, for example:

```
emb = transform(torch.nn.Embedding)(10, 8)
memo = set()
emb in memo
```
or also running the tests after adding `__eq__`. I think it is getting in conflict with other `__eq__` methods.

Reply:
quick solution to get around needing `__eq__`:
```
is_entire_transform = isinstance(self.pd_preprocess, Identity) and isinstance(self.pd_postprocess, Identity)
```